### PR TITLE
CI: add diff-shades integration

### PR DESCRIPTION
**Pull request by ichard26**
### Description

After a few months worth of work, [diff-shades](https://github.com/ichard26/diff-shades) and the psf/black integration is now ready to be reviewed! If you've living under a rock then just take [one look at this comment on the mypy repository](https://github.com/python/mypy/pull/11717#issuecomment-992275984) -- that's what this PR aims to add to our project!

The main reason why this is so complex is that the workflow supports three modes of operation, two of them exist to resolve our needs defined in psf/black#1390. The other one is a workflow_dispatch, i.e. basically you can run diff-shades against arbitrary revisions/branches/tags (including the latest PyPI release and a PR!).

diff-shades-dispatch

I plan on adding more documentation about this integration eventually but I want to get this landed sooner than later right now.

I really really hope this doesn't break in production but I guess there's only one way to tell :)

Resolves GH-1390.

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary? -> as cool as this is, it's still development only
- [x] Add / update tests if necessary? -> well I tested this manually a bunch
- [x] Add new / update outdated documentation?

### Review notes

This is 500+ lines of code and documentation, I don't expect you to review each line. Your main goal should be to point out as many spots you go "huh why is that there" so if **you** ever have to maintain / touch this you aren't as lost :slightly_smiling_face: 

Example runs:

- PR: when there's no changes: https://github.com/ichard26/black/pull/4#issuecomment-999826235
- PR: when there _are_ changes: https://github.com/ichard26/black/pull/5#issuecomment-999845088
- PR: when a file failed to format: https://github.com/ichard26/black/pull/6#issuecomment-1000473697
- Push to `main`: https://github.com/ichard26/black/actions/runs/1674523335
- Manually triggered: https://github.com/ichard26/black/actions/runs/1674718329

Documentation preview: https://ichard26-testblackdocs.readthedocs.io/en/main/contributing/gauging_changes.html#diff-shades

### CODE REVIEW

**Comment by ichard26** [comment]
1. primer is now failing, seriously? haha
2. sadly we can only test the PR comment functionality after this is merged -- [see note bullet point one here](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#workflow_run) 
3. I need to tweak the verbosity of the analyze steps cause it's wayyy too high right now

**Comment by ichard26** [comment]
Hmm there's some extra features in diff-shades I want to implement before landing this:

- [ ] Support for configuring project-specific custom arguments - this would be useful to temporarily ignore failing files / stress-test ESP
- [ ] Jupyter support

**Comment by onerandomusername** [comment]
> Hmm there's some extra features in diff-shades I want to implement before landing this:
> 
>     * [ ]  Support for configuring project-specific custom arguments - this would be useful to temporarily ignore failing files / stress-test ESP
> 
>     * [ ]  Jupyter support

maybe mark the pull as a draft so it doesn't get merged without these?

**Comment by ichard26** [comment]
No one is merging this other than me for at least a few days, I'm pretty much the only active write+ maintainer right now :p

**Comment by ichard26** [comment]
Actually, there's some more stuff that needs discussion / work.

File: docs/contributing/gauging_changes.md
Comment (JelleZijlstra) [inline docs/contributing/gauging_changes.md:72]:
```suggestion
Once finished, a PR comment will be posted embedding a summary of the changes and links to
```

**Comment by ichard26** [comment]
OK, this PR is ready for review. I've deferred the two items above as they aren't strictly necessary and will need much more care / effort to add to diff-shades than I'd like to hold this PR for. The main consideration is how we want to handle the distinction between Black w/ `--preview` and Black without. If #2752 gets merged soon, I'd prefer making this setup / job with the PR comment functionality the one with `--preview` enabled. If we want to add a job to assert that formatting doesn't change I'd probably implement it this way:

- Runs only on PRs (and _maybe_ main too)
- It doesn't post a PR comment but instead fails the workflow 
- I'd probably turn this workflow into an reusable workflow and then implement this workflow as a custom ref workflow_call (taking advantage of the workflow_dispatch support 🙂 ) to keep code reuse high

Thoughts?

---

Also, core team members please let me know if you'd like the commit bit for ichard26/diff-shades! I don't want be the sole maintainer / point of failure and [I'd be happy to see this list grow](https://github.com/ichard26/diff-shades#acknowledgements) :)

File: docs/contributing/gauging_changes.md
Comment (JelleZijlstra) [inline docs/contributing/gauging_changes.md:84]:
```suggestion
format. Otherwise a failure indicates a bug of the workflow.
```

**Review by JelleZijlstra — APPROVED** [review]

File: .pre-commit-config.yaml
Comment (JelleZijlstra) [reply .pre-commit-config.yaml:60]: We deleted it

Comment (ichard26) [reply .pre-commit-config.yaml:60]: Doesn't exist after #2717.

File: .pre-commit-config.yaml
Comment (ichard26) [reply .pre-commit-config.yaml:60]: sure, PR it

File: .github/workflows/diff_shades.yml
Comment (ichard26) [reply .github/workflows/diff_shades.yml:36]: The workflow checkouts older revisions of black and it's not known until later which revisions are needed

File: .github/workflows/diff_shades.yml
Comment (ichard26) [reply .github/workflows/diff_shades.yml:130]: Even if while running black over files a file does fail, we still want the diff-shades-comment workflow to fire and work successfully, that can't happen if this step fails skipping the PR body upload step.

File: scripts/diff_shades_gha_helper.py
Comment (ichard26) [reply scripts/diff_shades_gha_helper.py:272]: Not sure how this matters

File: .github/workflows/diff_shades.yml
Comment (onerandomusername) [reply .github/workflows/diff_shades.yml:130]: `if: always()` will make a step run no matter the previous outcomes

**Pull request merged**
